packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts

```diff
@@ -57,12 +67,22 @@ APPEND_OPCODES.add(Op.Primitive, (vm, { op1: primitive }) => {
 
 APPEND_OPCODES.add(Op.PrimitiveReference, vm => {
   let stack = vm.stack;
-  stack.pushJs(PrimitiveReference.create(check(stack.pop(), CheckPrimitive)));
-});
+  let value = check(stack.pop(), CheckPrimitive);
+  let ref;
+
+  if (value === undefined) {
+    ref = UNDEFINED_REFERENCE;
+  } else if (value === null) {
+    ref = NULL_REFERENCE;
+  } else if (value === true) {
+    ref = TRUE_REFERENCE;
+  } else if (value === false) {
+    ref = FALSE_REFERENCE;
+  } else {
+    ref = createPrimitiveRef(value);
+  }
```

Comment:
Seems like this is precisely what `switch` is for?

Reply:
@krisselden has said we should avoid using `switch` with non-numbers, because it's not very efficient.

Reply:
O_o ooooookkkkkiiiiieeeeeeeee ddoiooooooookkkkkiiiieeee